### PR TITLE
A module for shutting down tasks easily

### PR DESCRIPTION
**Pull request by mystenmark**
The motivation for this is that we would like to forcibly cancel all ActiveAuthority processes during reconfiguration. Currently, it is possible for active processes to observe intermediate states during reconfiguration (for example, the committee and AuthorityAggregator are updated non-atomically).

Additionally, this should make it possible to use a destroy-and-rebuild pattern for active processes, which is much simpler and less error-prone than the current approach of reconfiguring processes on the fly.

The design intentionally avoids keeping track of every JoinHandle ever created and calling `abort()` on them.

Also, it currently only supports tasks that do not return a value. We can add that later if necessary.

### CODE REVIEW

**Comment by bmwill** [comment]
> Additionally, this should make it possible to use a destroy-and-rebuild pattern for active processes, which is much simpler and less error-prone than the current approach of reconfiguring processes on the fly.

This is definitely going to be a situational thing. There are definitely situations where destroy and rebuild is going to be much more challenging than reconfiguration a process on the fly. Specifically any component that binds on a network interface I'll push strongly for not going the destroy-and-rebuild path.

File: crates/sui-core/src/authority_active/cancellable_spawner.rs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+use tap::TapFallible;
+
+use sui_types::error::{SuiError, SuiResult};
+
+/// CancellableSpawner allows spawned tasks to be cancelled easily without keeping track of every
+/// JoinHandle that gets created.
+pub struct CancellableSpawner {
+    cancel_tx: broadcast::Sender<()>,
+    spawn_lock: Arc<RwLock<()>>,
+    task_lock: Arc<RwLock<()>>,
```
Comment (bmwill) [inline crates/sui-core/src/authority_active/cancellable_spawner.rs:19]: I do not like using Locks in this fashion, where they guard nothing. For the spawn lock i would probably create some inner struct that has the functionality to actually do the spawning so that the functionality is truly gated by the lock

File: crates/sui-core/src/authority_active/cancellable_spawner.rs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+        debug!("sending cancel broadcast");
+        let _ = self.cancel_tx.send(());
+
+        // await all tasks exiting.
+        let _ = self.task_lock.write().await;
+        debug!("all tasks exited");
+
+        CancelGuard(spawn_guard)
+    }
```
Comment (bmwill) [inline crates/sui-core/src/authority_active/cancellable_spawner.rs:84]: Just be aware that in the case the inner task has spawned something onto the runtime, not via this mechanism, then there won't be anything to cancel those child tasks.

File: crates/sui-core/src/authority_active/cancellable_spawner.rs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        let _spawn_guard = self
+            .spawn_lock
+            .try_read()
+            .tap_err(|_| {
+                debug!("task could not be spawned, CancellableSpawner::cancel() is in progress")
+            })
+            .map_err(|_| SuiError::TaskSpawnError)?;
+
+        // While spawn_lock is held, task_lock.write() cannot be called, so try_read must succeed.
+        let task_guard = self.task_lock.clone().try_read_owned().unwrap();
```
Comment (bmwill) [inline crates/sui-core/src/authority_active/cancellable_spawner.rs:51]: I think i would probably avoid using a lock, because i think its error prone, and instead allocate some object that can be used as a notification once a particular task has completed. I mean you already have access to the JoinHandle from the tokio::spawn call, why not track those and call abort on those explicitly? (you can then await the join handles if you really want to wait for drop logic to run)

File: crates/sui-macros/src/lib.rs [hunk 4]
```diff
@@ -28,35 +28,39 @@ pub fn init_static_initializers(_args: TokenStream, item: TokenStream) -> TokenS
...
+                                )
+                                .start(::sui_simulator::anemo::Router::new())
+                                .unwrap()
+                        };
+                        let n1 = make_network(80);
+                        let n2 = make_network(81);
+
+                        let _peer = n1.connect(n2.local_addr()).await.unwrap();
+                    });
+                }
```
Comment (bmwill) [inline crates/sui-macros/src/lib.rs:63]: Is this an unrelated change?

File: crates/sui-core/src/authority_active/cancellable_spawner.rs (hunk 2 above)
Comment (mystenmark) [reply crates/sui-core/src/authority_active/cancellable_spawner.rs:84]: of course - its assumed that you'll have to use this for all spawning if you want your tasks cleaned up.

File: crates/sui-core/src/authority_active/cancellable_spawner.rs (hunk 3 above)
Comment (mystenmark) [reply crates/sui-core/src/authority_active/cancellable_spawner.rs:51]: the reason i didn't keep the join handles around to call abort() on is that then I would have to have a container to hold the handles, and then write additional code to clean up any handles that finished on their own. (Or else let the container grow without bound).

That's not insurmountable, but this seemed like less code, simpler code, and thus less error prone.

Can you explain why this technique might cause problems? (beyond just saying that locks are hard?)

File: crates/sui-macros/src/lib.rs (hunk 4 above)
Comment (mystenmark) [reply crates/sui-macros/src/lib.rs:63]: yes, I have no idea how this got into this branch! its not even a stray commit.

**Comment by mystenmark** [comment]
> This is definitely going to be a situational thing. There are definitely situations where destroy and rebuild is going to be much more challenging than reconfiguration a process on the fly. Specifically any component that binds on a network interface I'll push strongly for not going the destroy-and-rebuild path.

Ok - I'm not planning to use this in any situations that meet that description. This is intended for the active processes owned by ActiveAuthority, none of which are servers.

File: crates/sui-core/src/authority_active/cancellable_spawner.rs (hunk 1 above)
Comment (mystenmark) [reply crates/sui-core/src/authority_active/cancellable_spawner.rs]: Refactored to move the task spawning inside the spawn lock. I find the new version of the code much less clear but it does at least make lock order mistakes impossible since the nesting enforces the lock order.

There's still nothing to put inside the innermost lock, but that's because its really just a more convenient alternative to an atomic task counter + a condvar.
